nixos/promscale: init

0x4A6F opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Enable service for [timescale-prometheus](https://github.com/timescale/timescale-prometheus).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

0x4A6F committed:
nixos/promscale: init

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,262 @@
+        '';
+      };
+
+      db-password = mkOption {
```
I'm not a fan of passwords in the nix store. We shouldn't do this. :-1:

0x4A6F replied:
Neither am I. Is this implementation sufficient or have I missed something from passwordFile best practice?

aanderse replied:
Unfortunately your new `db-passwordFile` option still ends up in the nix store.

0x4A6F replied:
Thanks for pointing this out.
I wonder if my head is now able to come up with a solution.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        '';
+      };
+
+      db-connect-retries = mkOption {
```
Again, purely a style thing, feel free to ignore 100% and then some but... I think most modules typically use a `database.` prefix for database related options, something like
```
database = {
  connect-retries = mkOption { ... };
  host = mkOption { ... };
  name = mkOption { ... };
};
```

Consistency is always nice.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        description = "The timescale-prometheus package to use.";
+      };
+
+      async-acks = mkOption {
```
Purely a style thing, feel free to ignore, but I'm curious why you chose `-` instead of camel case when almost every NixOS module uses camel case.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+      };
+
+      db-ssl-mode = mkOption {
+        type = types.str;
```
```suggestion
        type = types.enum [ "disable" "allow" "prefer" "require" "verify-ca" "verify-full" ];
```

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        type = types.str;
+        default = "localhost";
+        description = ''
+          The TimescaleDB host [TS_PROM_DB_HOST] (default "localhost")
```
the `(default "localhost")` text is redundant as you have already provided a `default` value.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+      };
+
+      log-level = mkOption {
+        type = types.nullOr types.str;
```
```suggestion
        type = with types; nullOr (enum [ "error" "warn" "info" "debug" ]);
```

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+      };
+
+      migrate = mkOption {
+        type = types.str;
```
Why are you not using a `bool` here?

aanderse replied:
You should use `lib.booleanToString`, or whatever the function that converts bools to their textual representation then.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+      migrate = mkOption {
+        type = types.str;
+        default = "true";
+        example = "false";
```
**If** this is actually a `bool` then `example` provides no value here.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        '';
+      };
+
+      web-listen-address = mkOption {
```
I would add an example here specifying both a host/ip and port.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+          [ "-async-acks"
+          ]
+        '';
+        description = ''
```
Is there upstream documentation on possible values we can link to?

aanderse replied:
Sounds good :+1:

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+      after = [ "postgresql.service" ];
+
+      environment = {
+        TS_PROM_ASYNC_ACKS = "${toString cfg.async-acks}";
```
I could be entirely wrong here but I assume values under `environment` automatically cast to string anyways. Is that not the case?

Additionally quoting is not required:
```
TS_PROM_ASYNC_ACKS = toString cfg.async-acks;

# or if toString is not required
TS_PROM_ASYNC_ACKS = cfg.async-acks;
```

aanderse replied:
Now I know. Thanks.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        TS_PROM_WEB_TELEMETRY_PATH = "${toString cfg.web-telemetry-path}";
+      };
+
+      script = "
```
Why not just use `serviceConfig.ExecStart`?

0x4A6F replied:
`script` usage:
```
exec ${cfg.package}/bin/example \
  ${optionalString (cfg.passwordFile != null) "--password $(<${cfg.passwordFile})"}
```
- [x] no leak to store
- [x] no leak to process list (ps fauxwww | grep password)

---

`serviceConfig.ExecStart` usage:
```
${cfg.package}/bin/example \
  ${optionalString (cfg.passwordFile != null) "--password $(head -n 1 ${cfg.passwordFile})"}
```
- [x] no leak to store
- [?] no leak to process list (ps fauxwww | grep password)

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        PrivateDevices = true;
+        # Sandboxing
+        CapabilityBoundingSet = "CAP_NET_RAW CAP_NET_ADMIN";
+        ProtectSystem = "strict";
```
Implied by `DynamicUser`.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        CapabilityBoundingSet = "CAP_NET_RAW CAP_NET_ADMIN";
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        PrivateTmp = true;
```
Implied by `DynamicUser`.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
```
Implied by `DynamicUser`.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,272 @@
+        StateDirectoryMode = "0700";
+        PrivateDevices = true;
+        # Sandboxing
+        CapabilityBoundingSet = "CAP_NET_RAW CAP_NET_ADMIN";
```
Not overly important, and I don't know for sure, but I'm going to guess that a list is preferred here:
```suggestion
        CapabilityBoundingSet = [ "CAP_NET_RAW" "CAP_NET_ADMIN" ];
```

aanderse commented on nixos/tests/timescale-prometheus.nix:
```diff
@@ -0,0 +1,98 @@
+              settings = {
+                shared_preload_libraries = "timescaledb";
+              };
+              ensureDatabases = [ "${db-name}" ];
```
Quoting not required.

aanderse commented on nixos/tests/timescale-prometheus.nix:
```diff
@@ -0,0 +1,98 @@
+              ensureDatabases = [ "${db-name}" ];
+              ensureUsers = [
+                {
+                  name = "${db-user}";
```
Quoting not required.

aanderse reviewed (changes requested):
Thanks for taking care of the password issue!

I have taken the time to actually give this module a review now. I hope at least some of the comments are helpful.

0x4A6F reviewed (commented):
Thanks for your review.

aanderse commented:
Changes are looking good. I left a few minor comments.

I should have asked before but made the assumption it cannot... so I'll ask now: can this program accept a configuration file instead of environment variables?

0x4A6F commented on nixos/tests/timescale-prometheus.nix:
```diff
@@ -0,0 +1,155 @@
+#   -db-connections-max int
+#     	maximum connections that can be open at once, defaults to 80% of the max the DB can handle [TS_PROM_DB_CONNECTIONS_MAX] (default -1)
+#   -db-host string
+#     	The TimescaleDB host [TS_PROM_DB_HOST] (default "localhost")
+#   -db-name string
+#     	The TimescaleDB database [TS_PROM_DB_NAME] (default "timescale")
+#   -db-password string
+#     	The TimescaleDB password [TS_PROM_DB_PASSWORD]
+#   -db-port int
+#     	The TimescaleDB port [TS_PROM_DB_PORT] (default 5432)
+#   -db-ssl-mode string
+#     	The TimescaleDB connection ssl mode [TS_PROM_DB_SSL_MODE] (default "require")
+#   -db-user string
+#     	The TimescaleDB user [TS_PROM_DB_USER] (default "postgres")
+#   -db-writer-connection-concurrency int
+#     	maximum number of database connections per go process writing to the database [TS_PROM_DB_WRITER_CONNECTION_CONCURRENCY] (default 4)
+#   -labels-cache-size uint
+#     	maximum number of labels to cache [TS_PROM_LABELS_CACHE_SIZE] (default 10000)
+#   -leader-election-pg-advisory-lock-id int
+#     	Unique advisory lock id per adapter high-availability group. Set it if you want to use leader election implementation based on PostgreSQL advisory lock. [TS_PROM_LEADER_ELECTION_PG_ADVISORY_LOCK_ID]
+#   -leader-election-pg-advisory-lock-prometheus-timeout slack
+#     	Adapter will resign if there are no requests from Prometheus within a given timeout (0 means no timeout). Note: make sure that only one Prometheus instance talks to the adapter. Timeout value should be co-related with Prometheus scrape interval but add enough slack to prevent random flips. [TS_PROM_LEADER_ELECTION_PG_ADVISORY_LOCK_PROMETHEUS_TIMEOUT] (default -1ns)
+#   -leader-election-rest
+#     	Enable REST interface for the leader election [TS_PROM_LEADER_ELECTION_REST]
+#   -log-level string
+#     	The log level to use [ "error", "warn", "info", "debug" ]. [TS_PROM_LOG_LEVEL] (default "debug")
+#   -metrics-cache-size uint
+#     	maximum number of metric names to cache [TS_PROM_METRICS_CACHE_SIZE] (default 10000)
+#   -migrate string
+#     	Update the Prometheus SQL to the latest version. Valid options are: [true, false, only] [TS_PROM_MIGRATE] (default "true")
+#   -scheduled-election-interval duration
+#     	Interval at which scheduled election runs. This is used to select a leader and confirm that we still holding the advisory lock. [TS_PROM_SCHEDULED_ELECTION_INTERVAL] (default 5s)
+#   -tput-report int
+#     	interval in seconds at which throughput should be reported [TS_PROM_TPUT_REPORT]
+#   -web-cors-origin string
+#     	Regex for CORS origin. It is fully anchored. Example: 'https?://(domain1|domain2)\.com' [TS_PROM_WEB_CORS_ORIGIN] (default ".*")
+#   -web-listen-address string
+#     	Address to listen on for web endpoints. [TS_PROM_WEB_LISTEN_ADDRESS] (default ":9201")
+#   -web-telemetry-path string
+#     	Address to listen on for web endpoints. [TS_PROM_WEB_TELEMETRY_PATH] (default "/metrics")
```
Here are the command line arguments.
I see no ability to specify a configuration file.

Thanks again for the review. In the meantime I did some more work.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,294 @@
+
+      description = "Timescale-Prometheus connector";
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "postgresql.service" ];
```
Shouldn't this be conditional on whether db host is local?

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,294 @@
+      asyncAcks = mkOption {
+        type = types.nullOr types.bool;
+        default = null;
+        description = ''
```
Please be consistent with period at end of `description`.

aanderse commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,294 @@
+      }
+      {
+        assertion = (cfg.leaderElection.rest == true) -> (cfg.leaderElection.pg.advisoryLockId == null && cfg.leaderElection.pg.advisoryLockPrometheusTimeout == null);
+        message = "When \"services.timescale-prometheus.leaderElection.rest\" is set, following must *NOT* be set:\n\ \ \ \ \ - \"services.timescale-prometheus.leaderElection.pg.advisoryLockId\"\n\ \ \ \ \ - \"services.timescale-prometheus.leaderElection.pg.advisoryLockPrometheusTimeout\"";
```
What's with the `\` characters?

aanderse reviewed (changes requested):
This looks great. Do we have any subject matter experts in the community that can give this a quick one over? From my perspective we're pretty much ready for a merge :+1:

0x4A6F commented on nixos/modules/services/monitoring/timescale-prometheus.nix:
```diff
@@ -0,0 +1,294 @@
+      requires = [ "postgresql.service" ];
+      after = [ "postgresql.service" ];
+
+      environment = {
```
Some command line arguments are ignored, if `cfg.leaderElection.rest` is true.

0x4A6F commented:
Thanks, MR is updated addressing your suggestions.

I should also extend `services.prometheus` with [remote_write](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write) and [remote_read](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_read). Probably in another MR, when ready to merge this one.

aanderse reviewed (approved):
This looks fine to me but you need someone who knows more about `prometheus` to review this before merge please.

0x4A6F commented:
I've updated again to reflect the project rename to promscale.

Thanks for your review. TIme to search for our prometheus experts.

aanderse commented:
You might consider mentioning this on one of the discourse threads.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/324

andir commented on nixos/tests/promscale.nix:
```diff
@@ -0,0 +1,146 @@
+            # "-web-cors-origin example.com/public/.*" # corsOrigin:0xc00016dea0
+            ];
+            services.postgresql = {
+              authentication = pkgs.lib.mkOverride 10 ''
```
You can probably get rid of this by using `/run/postgresql/` as "host" in this test. By default postgresql uses the unix user that connects (on NixOS)

0x4A6F replied:
Doesn't seem to work without `authentication` set and `db-host = "/run/postgresql/";`.
Seems with these settings promscale is started before postgres is up.

0x4A6F commented:
Currently with promscale 0.1.1 (#102013) the test breaks, because of postgres version mismatch.
Now it needs postgresql 12, tested with following in my local branch:

```
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix

-  postgresql = postgresql_11.override { this = postgresql; };
+  postgresql = postgresql_12.override { this = postgresql; };
   postgresqlPackages = recurseIntoAttrs postgresql.pkgs;
   postgresql11Packages = pkgs.postgresqlPackages;
+  postgresql12Packages = pkgs.postgresqlPackages;
```

andir commented:
On 15:52 29.10.20, 0x4A6F wrote:
> @0x4A6F commented on this pull request.
> Doesn't seem to work without `authentication` set and `db-host = "/run/postgresql/";`.
> Seems with these settings promscale is started before postgres is up.

I got it working while trying to deploy it to a server. Unfortunately I
didn't continue that route as it was insisting on being superuser on my
postgres server.

JamesGuthrie commented:
@0x4A6F I am taking a look at using this with the latest version of promscale. Are you interested in continuing work on this change? I could support you in getting it merged if so.

benpye commented:
This looks possible now, as of 2.5.0 timescaledb supports PostgreSQL 14 and promscale 0.8.0 adds support. I've gotten this working in my flake - https://github.com/benpye/nix-config/blob/main/modules/services/promscale.nix . Do you want to try and update this PR or are you happy if I try and submit my version?
